code/go/0chain.net/core/common/lock.go

```diff
@@ -0,0 +1,78 @@
+package common
+
+import (
+	"sync"
+)
+
+type Lock struct {
+	key string
+	// stale signals if lock is deleted from map. if true, discard use of this lock and get fresh lock
+	stale      bool
+	actualLock *sync.Mutex
+	// count how many process requires this lock
+	count   int32
+	countMu *sync.Mutex
+
+	// Parent Map
+	pMap *MapLocker
+}
+
+// Lock Acquire lock
+func (l *Lock) Lock() {
+	for {
+		l.actualLock.Lock()
+		if l.stale {
+			l = l.pMap.GetLock(l.key)
```

Comment:
This assignment of `lock` will not work, it tries to change the receiver pointer. It does not work, just like we try to change the pointer of a passed in parameter of a method. So the result would be we still using the staled lock even we actually get the new one from pMap, while this staled lock was actually had been removed from pMap when the `Unlock()` was called, so using this stale lock is dangerous as it does not guarantee the thread safe of resources that it suppose to protect under the same key.

Reply:
This is severe issue. Thanks a lot for pointing it out.
If I can fix this issue, I suppose I can use this locker. Do you suggest so?